Packs/knowbe4Phisher/Integrations/knowbe4Phisher/knowbe4Phisher.py

```diff
@@ -0,0 +1,812 @@
+
+    argument: a multiline string with a request in a human readable format
+
+    return: a final string that is formatted to suit the http_request function
+    """
+    format_desc = st.splitlines()
+    final = ""
+    for i in format_desc:
+        final = final + i + '\\n'
+    final = final[:-2]
+    return final
+
+
+def create_gql_query(payload: str, query: str):
+    """
+    for gql query - the query argument inside the request should be with double quotes.
+    this function is inserting the query parameter to the request body and adds the needed parameters,
+    for gql query - wrapping the query in a dictionary representation
+
+    arguments: payload, the body part of a request. query, the query parameter from phisherMessages arguments
+
+    output: a string with dict representation of gql query.
+    """
+    st = "query:"
+    j = len(st)
+    i = payload.find(st)
+    temp = payload[:i + j] + query + payload[i + j:]
+    return "{\"query\": \"" + temp + "\",\"variables\": {}}"
+
+
+def get_last_fetch_time(last_run: dict, first_fetch_time: str):
+    """
+    this helper function will get the demisto.getLastRun() and first_fetch_time parameter
+    and will return:
+    last_time - timestamp of last message fetched
+    query - the query parameter with the timestamp to match the Phisher filter
+    """
+    # Get the last fetch time, if exists
+    utc = pytz.UTC
+    last_fetch = last_run.get('last_fetch')
+    # Handle first time fetch
+    if last_fetch is None:
+        last_fetch = dateparser.parse(first_fetch_time)
+        time_param = last_fetch.strftime(DATE_FORMAT)
+        query = "\\\" reported_at:[" + time_param + " TO *]\\\""
+    else:
+        last_fetch = dateparser.parse(last_fetch)
+        time_param = last_fetch.strftime(DATE_FORMAT)
+        query = "\\\" reported_at:[" + time_param + " TO *]\\\""
+    last_time = last_fetch.replace(tzinfo=utc)
+    return last_time, query
+
+
+def calculate_number_of_events(client: Client, query: str):
+    """
+    function gets the formatted query with the relevant timestamp
+    and returns number of events to fetch
+    """
+    # get number of new events
+    payload = create_gql_query(NUMBER_OF_MESSAGES, query)
+    req = create_request(payload)
+    limit_req = client.phisher_get_number_of_messages_request(req)
+    return str(limit_req.get('data', {}).get('phisherMessages', {}).get('pagination', {}).get('totalCount'))
+
+
+''' COMMAND FUNCTIONS '''
+
+
+def test_module(client: Client, params: dict) -> str:
+    """Tests API connectivity, authentication and parameters for fetch incidents
+
+    Returning 'ok' indicates that the integration works like it is supposed to.
+    Connection to the service is successful.
+    Raises exceptions if something goes wrong.
+
+    param Client: client to use
+
+    return: 'ok' if test passed, specific error if with a suggestion of how to fix.
+    """
+    first_fetch_time = params.get('first_fetch')
```

Comment:
i added the max_fetch and first_fetch to the client.

i don't agree with the suggestion - i have a function that deals with dates, so it doesn't make sense to change first_fetch everywhere, instead of changing it once in a function

Reply:
fixed according to suggestion